utils/formatDate.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * Utility file to handle datetimes.
+ *
+ * Because this is a State of California application, we assume all times are
+ * in Pacific Time if there is no timezone provided in the datetime string.
+ */
+import { format, isValid, parse } from 'date-fns'
+import { zonedTimeToUtc } from 'date-fns-tz'
+import { ApiGatewayDateString } from '../types/common'
+
+const pacificTimeZone = 'America/Los_Angeles'
+const apiGatewayFormat = "yyyy-MM-dd'T'HH:mm:ss"
+
+/**
+ * Parse a date string from the API gateway.
+ */
+export function parseApiGatewayDate(dateString: ApiGatewayDateString): Date {
+  return parse(dateString, apiGatewayFormat, new Date())
+}
+
+/**
+ * Assume input is in Pacific Time and convert to UTC.
+ */
+export function datetimeInUtc(date: Date | number | string): Date {
+  return zonedTimeToUtc(date, pacificTimeZone)
+}
+
+/**
+ * Parse and convert an API gateway string to UTC.
+ */
+export function parseConvertDate(dateString: ApiGatewayDateString): Date {
+  const parsedDate = parseApiGatewayDate(dateString)
+  return datetimeInUtc(parsedDate)
+}
+
+/**
+ * Return a string that matches the API gateway format for datetimes.
+ */
+export function formatFromApiGateway(date: Date): string {
+  return format(date, apiGatewayFormat)
+}
+
+/**
+ * Create a Date object that is offset from today.
+ */
+export function getDateWithOffset(offset = 1): Date {
+  const today = new Date()
+  const sometime = today.setDate(today.getDate() + offset)
+  const sometimeUtc = datetimeInUtc(sometime)
+  return sometimeUtc
```

Comment:
I was thinking that it would be safer to have everything in UTC, but now that I'm actively working on the time slot parsing, I'm realizing that it might be the wrong choice. I'll revise this.

Reply:
Ok after some more work, I've refactored a number of things and added _a lot more_ comments to `formatDate.ts`, reproduced here for reference.

According to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date:
> A JavaScript date is fundamentally specified as the number of milliseconds that have
> elapsed since midnight on January 1, 1970, UTC. ... It's important to keep in mind
> that while the time value at the heart of a Date object is UTC, the basic methods to
> fetch the date and time or its components all work in the local (i.e. host system)
> time zone and offset.

So by default, if a date time string has no time zone (which is how we expect the date times from the API gateway to be formatted), it is interpreted as a UTC time, but all the fetch and format functions render the date in the host system's time zone.

From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse:
> For example, "2011-10-10" (date-only form), "2011-10-10T14:48:00" (date-time form),
> or "2011-10-10T14:48:00.000+09:00" (date-time form with milliseconds and time zone)
> can be passed and will be parsed. When the time zone offset is absent, date-only forms
> are interpreted as a UTC time and date-time forms are interpreted as local time.

I've refactored the code a bit to ensure that we're aligning with the underlying UTC storage format. `parseApiGatewayDate()` now looks like this:

```js
export function parseApiGatewayDate(dateString: ApiGatewayDateString): Date {
  return toDate(dateString, { timeZone: pacificTimeZone })
}
```

It takes advantage of `date-fns-tz`'s [`toDate()`](https://github.com/marnusw/date-fns-tz#todate) functionality which allows us to specify the originating time zone for a date time string that does not specify the time zone.

Example from `date-fns-tz`'s docs:
```js
// When there is no offset in the date string the timeZone property is used
const bangkokDate = toDate('2014-10-25T13:46:20', { timeZone: 'Asia/Bangkok' })
```

When there is an offset in the date time string, the `timeZone` option is ignored:
```js
// Since toDate simply clones a Date instance timeZone option is effectively ignored in this case
const date = new Date('2014-10-25T13:46:20Z')
const clonedDate = toDate(date, { timeZone: 'Europe/Paris' })
```